llama-index-integrations/readers/llama-index-readers-microsoft-sharepoint/llama_index/readers/microsoft_sharepoint/base.py

```diff
@@ -131,6 +131,40 @@ def _get_access_token(self) -> str:
             logger.error("Error retrieving access token: %s", json_response["error"])
             raise ValueError(f"Error retrieving access token: {error_message}")
 
+    def _send_request_with_retry(self, request: requests.Request) -> requests.Response:
+        """
+        Makes a request to the SharePoint API with the provided request object.
+        If the request fails with a 401 status code, the access token is refreshed and the request is retried once.
+        """
+        curr_headers = (request.headers or {}).copy()
+        curr_headers.update(self._authorization_headers)
+        request.headers = curr_headers
+        prepared_request = request.prepare()
+        with requests.Session() as session:
+            response = session.send(prepared_request)
+
+            if response.status_code == 401:
+                # refresh token
+                logger.debug("Received 401. Refreshing access token.")
+                access_token = self._get_access_token()
+                self._authorization_headers = {
+                    "Authorization": f"Bearer {access_token}"
+                }
+                curr_headers.update(self._authorization_headers)
+                request.headers = curr_headers
+                prepared_request = request.prepare()
+                response = session.send(prepared_request)
+
+            response.raise_for_status()
+            return response
+
+    def _send_get_with_retry(self, url: str) -> requests.Response:
```

Comment:
Should we be using tenacity for this?

Reply:
(I guess not for something this specific?)

Reply:
Hmm not sure it really makes sense to use tenacity when we only need to retry the request once